### ISSUE TITLE
Physical constraint problem on tang_primer_20k

### ISSUE THREAD

**Issue by mmx-114514** [enjoy-digital/litex#1820]
When I try to enable both hdmi and etnerbone on tang_primer_20k, the generated .cst physical constraint file has inconsistent voltages in the same bank: Bank 0 vccio(3.3) is locked by other constraint or embedded port, conflicting BANK_VCCIO set by 'hdmi_clk_p_obuf' : IO_TYPE = LVCMOS18 in the same bank. port, conflicting BANK_VCCIO set by 'hdmi_clk_p_obuf' : IO_TYPE = LVCMOS18 in the same bank How to solve this? Thank you for your help.

**Comment by hvegh** [comment]
Yes... all the hdmi pins in the dock are pulled up using an external resistor:

Drop the  IOStandard("LVCMOS18") on these pins:

        Subsignal("hdp", Pins("CARD1:154"), IOStandard("LVCMOS18")),
        Subsignal("cec", Pins("CARD1:152"), IOStandard("LVCMOS18")),
        Subsignal("sda", Pins("CARD1:95"),  IOStandard("LVCMOS18")),
        Subsignal("scl", Pins("CARD1:97"),  IOStandard("LVCMOS18")),

**Comment by enjoy-digital** [comment]
Hi @mmx-114514,

Can you share your build command to reproduce the issue? Thanks.

**Comment by mmx-114514** [comment]
I used the way in linux-on-litex-vexriscv to generate vexriscv-smp via make.py, I used "python3 . /make.py --board=Sipeed_tang_primer_20k --with-coherent-dma --build" to generate the coherent DMA, and also added framebuffer and with video terminal. currently my screen displays the terminal information correctly. But I can find by coremark that the cpu score is much lower than without hdmi, I think this is caused by not being able to call video dma. Also in the generated dts file I can't find any configuration related to video with hdmi.
Thanks.

**Comment by mmx-114514** [comment]
I Drop the IOStandard("LVCMOS18") on these pins.
    Subsignal("hdp", Pins("CARD1:154"), #IOStandard("LVCMOS18")),
    Subsignal("cec", Pins("CARD1:152"), #IOStandard("LVCMOS18")),
    Subsignal("sda", Pins("CARD1:95"),  #IOStandard("LVCMOS18")),
    Subsignal("scl", Pins("CARD1:97"),  #IOStandard("LVCMOS18")),

**Comment by enjoy-digital** [comment]
This should be fixed now with:
  - enjoy-digital/litex@43733f2556103a4c2d33533e921b42423bcee287
  - litex-hub/litex-boards@708d8a80680bd63337ae1f30d30672a7f562f70d

The issue was twofold: the Tang Primer 20K HDMI TMDS pins were missing explicit LVCMOS33 constraints, and the Gowin CST backend was treating any `_p`/`_n` resource name as a differential pair.

LiteX now keeps explicitly single-ended LVCMOS `_p`/`_n` pins as separate CST constraints, while preserving the previous pairing behavior for unspecified or explicitly differential IO standards. LiteX- Boards now also marks the Tang Primer 20K HDMI TMDS pins as LVCMOS33.

I re-tested the reported build:

      ./sipeed_tang_primer_20k.py --with-video-terminal --with-etherbone --build

With current LiteX/LiteX-Boards, the Gowin flow completes through bitstream generation, so I think this can be closed.
